### PR TITLE
perf(ext/fetch): use content-length in InnerBody.consume

### PR DESCRIPTION
**Pull request by marcosc90**
For requests where `content-length` header is present, we can allocate the buffer before reading the whole response reducing memory usage and increasing performance since there's no need for pushing into an array and looping twice through the chunks.

---

Take the following script:

```
truncate -s 2g /tmp/out.bin
```
**server**
```js
const server = Deno.listen({ port: 3031 });
const file = await Deno.readFile('/tmp/out.bin')

for await (const conn of server) {
  serveHttp(conn);
}

async function serveHttp(conn) {
  const httpConn = Deno.serveHttp(conn);
  for await (const requestEvent of httpConn) {
    requestEvent.respondWith(
      new Response(file, {
        status: 200,
        headers: new Headers({ 'content-length': file.byteLength }) 
      }),
    );
  }
}
```

**fetch**
```js
// import { fetch } from 'undici'
const response = await fetch("http://localhost:3031");
const res = await response.arrayBuffer();

console.log(res.byteLength)
```

----

### Results

**Deno** / 1.25.3 (release, x86_64-unknown-linux-gnu)
```
User time (seconds): 7.16
System time (seconds): 1.44
Percent of CPU this job got: 211%
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:04.07
Maximum resident set size (kbytes): 4375544
```

**Node / Undici**

```
User time (seconds): 1.20
System time (seconds): 1.35
Percent of CPU this job got: 94%
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.72
Maximum resident set size (kbytes): 4287288
```

**This patch** (Release build)

```
User time (seconds): 1.35
System time (seconds): 0.90
Percent of CPU this job got: 79%
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.84
Maximum resident set size (kbytes): 2163648
```

Measured with: `/usr/bin/time --verbose {command}`

### CODE REVIEW

File: ext/fetch/22_body.js [hunk 1]
```diff
@@ -64,17 +64,21 @@
...
+    constructor(stream, { contentLength } = {}) {
       /** @type {ReadableStream<Uint8Array> | { body: Uint8Array | string, consumed: boolean }} */
       this.streamOrStatic = stream ??
         { body: new Uint8Array(), consumed: false };
       /** @type {null | Uint8Array | string | Blob | FormData} */
       this.source = null;
       /** @type {null | number} */
       this.length = null;
+
+      this.#contentLength = contentLength;
```
Comment (marcosc90) [inline ext/fetch/22_body.js:81]: Not sure yet what `this.length` is for, maybe it can be used instead of `this.#contentLength`

File: ext/fetch/22_body.js [hunk 2]
```diff
@@ -64,17 +64,21 @@
   }
 
   class InnerBody {
+    #contentLength = null;
+
     /**
      * @param {ReadableStream<Uint8Array> | { body: Uint8Array | string, consumed: boolean }} stream
      */
-    constructor(stream) {
+    constructor(stream, { contentLength } = {}) {
```
Comment (littledivy) [inline ext/fetch/22_body.js:72]: nit: avoid objects (perf reasons) since this is internal API
```suggestion
    constructor(stream, contentLength) {
```

**Review by littledivy — COMMENTED** [review]
Very cool!

File: ext/fetch/26_fetch.js [hunk 3]
```diff
@@ -333,8 +335,18 @@
         response.body = null;
         core.close(resp.responseRid);
       } else {
+        const contentLengthHeader = ArrayPrototypeFind(
+          resp.headers,
+          (entry) => byteLowerCase(entry[0]) === "content-length",
+        );
```
Comment (littledivy) [inline ext/fetch/26_fetch.js:341]: Faster way to do this that will have lesser impact on non-content-length header situations -

We already iterate over headers in the Rust op (`op_fetch_send`). Do this check in Rust and return `contentLength` value in the result.
 https://github.com/denoland/deno/blob/24f7f3fda984e7226b5858085a7ed1c53ab081e0/ext/fetch/lib.rs#L390-L392

Comment (marcosc90) [reply ext/fetch/26_fetch.js:341]: Don't know any rust tbh. I could give it a try.

Although we are already looping through headers on javascript side to match `location`:

https://github.com/denoland/deno/blob/65d2393c02a4ab637dd59149eb337631046346cc/ext/fetch/26_fetch.js#L371-L374

PS: different path, we're not looping twice.

**Review by lucacasonato — COMMENTED** [review]
We can not trust the `content-length` header. This code needs to be written in a way that `content-length` can be used as a size indication as a performance optimization, but the final size of the buffer is still determined by the size and number of actual chunks.

Specifically we need to deal with the cases:
- totalLength > `content-length`
- totalLength < `content-length`

**Comment by marcosc90** [comment]
> We can not trust the `content-length` header. This code needs to be written in a way that `content-length` can be used as a size indication as a performance optimization, but the final size of the buffer is still determined by the size and number of actual chunks.
> 
> Specifically we need to deal with the cases:
> 
> * totalLength > `content-length`
> * totalLength < `content-length`

Undici, Chrome, Firefox use `content-length` header, the size of the buffer is determined by the `content-length` header in all three. One can argue that if `content-length` header does not contain the right value, the response is corrupt.
And in those cases the user can still consume the body by iterating through the response stream. 

It's too big of optimization to ignore it.

**Comment by marcosc90** [comment]
If we want to behave differently, we can grow the buffer if more bytes are read and slice the buffer if less bytes are read, at a performance penalty for cases where `content-length != totalLength`.

FYI, if `content-length` > totalLength, `Deno` currently hangs. (same as the other runtimes/libs)

**Comment by lucacasonato** [comment]
I am confused by your comment. I am **not** suggesting we do not do this optimization. I am suggesting that your current implementation is placing too much trust in the `content-length` header, and that the implementation needs to be able to handle when the real body size deviates from the "known, well-formed, and correct `content-length`" fast path. There needs to be a way for the fast path to downgrade to the slow path if it detects that some assumptions made for the fast path do not hold anymore.

My concern here is not about the HTTP transport decoding which in HTTP/1 needs to place trust in the `content-length` header like you describe. `hyper` handles this for us, and I trust it to be correct. What I am worried about us re-interpreting these headers _differently_ to the transport layer, which can lead to issues.

For example, if a server returns a HTTP response (on a http/1 connection) containing both a `transfer-encoding: chunked` header, and a `content-length` header, Hyper may use the `transfer-encoding` header and ignore the `content-length` when decoding the stream. In the process it will return more data than the `content-length` indicates. On the JS side however we trust that the `content-length` is correct and just blindly assume that `hyper` returns the exact number of bytes specified in the `content-length` header. These are the scenarios where `content-length != totalLength`.

This same scenario described above is especially prevalent on HTTP/2 and HTTP/3, where body framing happens completely independently from headers such as `content-length`, so they have explicit meaning to the transport layer at all there.

Also instead of manually re-interpreting the `content-length` header ourselves, we should use use the value returned from `reqwest::Response::content_length()`. This value will automatically also take into account response body compression (e.g. `transfer-encoding: gzip`).

**Comment by lucacasonato** [comment]
If you need guidance on the Rust side changes for this, I'd be happy to help out! :)

**Comment by marcosc90** [comment]
Valid points. I meant to open the PR as a draft, to showcase the performance improvements. I'll add the code for downgrading to the slow path in case there's more data than reported by `content-length` header, as well as ignoring it when `transfer-encoding: chunked` is present as per the spec

https://greenbytes.de/tech/webdav/rfc2616.html#rfc.section.4.4
> If a message is received with both a Transfer-Encoding header field and a Content-Length header field, the latter must be ignored.

It'll take me a while to write any good rust code, so if someone wants to pitch in with adding `reqwest::Response::content_length()` as mentioned, I can handle the JS stuff.

**Comment by lucacasonato** [comment]
@marcosc90 I have wired through the `content-length` from the Rust side into the `InnerBody` constructor now :)

**Comment by marcosc90** [comment]
Oh, that was easy :P, I really need to start learning Rust. Thanks @lucacasonato, I'll start working on the slow path.

**Comment by marcosc90** [comment]
@lucacasonato Deno **debug** crashes if `content-length != totalLength`

```
thread 'tokio-runtime-worker' panicked at 'payload claims content-length of 10240, custom content-length 
header claims 20480', 
/home/marcos/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.19/src/proto/h1/role.rs:675:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Using release build, it does not crash but it's also impossible to test the downgrade since it seems that Hyper stops reading after `content-length` bytes. If I send more data than `content-length` that extra data is clipped.
If `content-length` is greater, the request never ends, it keeps waiting for the extra data that will never be sent. (The only thing I could test in this scenario is the request aborting by using an abort controller after N milliseconds)

I added a test, with `ignore: true` in the meantime. I performed manual tests by hardcoding `this.#contentLength`. We can enable the test if there's a way to mock `this.#contentLength` or `res.content_length` in rust.

Here's the test: https://github.com/denoland/deno/pull/15925/commits/5f9b57ab2b952e8282718310d792ef6f7728b28c if you have any ideas on how to enable it.

tl;dr: It seems that the downgrade scenario will never happen with current Hyper implementation.

**Comment by marcosc90** [comment]
> You can not have the test use a real hyper webserver as the server side

@lucacasonato The server is not the issue, the server actually sends back all the bytes, with the invalid content-type header. It's the client that's the issue.

Using `Deno.connect` I get all 10240 bytes, even if `Content-Length` is 5120
```
const write = new TextEncoder().encode(`GET / HTTP/1.1\r\n\r\n`);
const conn = await Deno.connect({ hostname: "127.0.0.1", port: 4514 })
await conn.write(write);
for await(const chunk of Deno.iter(conn)) {
	console.log(chunk); // 10240
}
```

Using `fetch` I only get `5120`.

**Comment by marcosc90** [comment]
After more testing:

- If `transfer-encoding` and `content-length` are present. `res.content_length()` is `null`.
- if `content-length`  > `totalLength`,  the request is never terminated, or the following error is thrown: ` Error: request or response body error: error reading a body from connection: end of file before message length reached`. (Depends on whether the server closes the request or not). 
- if `content-length`  <  `totalLength` Hyper reads a maximum of `content-length` bytes, the rest is discarded. In this case `res.content_length()` equals `content-length` header
- if several `content-length` headers are sent with different values, an error is triggered.

This makes the downgrade path impossible to occur with the current Hyper version since all cases are being handled correctly by Hyper (client).

We can leave the downgrade path as a safeguard (without unit tests, just manual testing), or we can remove it and trust Hyper.

**Comment by lucacasonato** [comment]
That is great. In that case, let's remove the downgrade path. Can you add a comment in the code with this axiom?

Also, let's rename `contentLength` to `knownExactLength` to make it clear this is not just a naive reading of the `content-length` header.

**Comment by lucacasonato** [comment]
Actually one more test: is `contentLength` always `null` for requests to HTTP/2 servers?

**Comment by marcosc90** [comment]
> Actually one more test: is `contentLength` always `null` for requests to HTTP/2 servers?

It is not. In the following test, `#knownExactLength` is 8, which is correct.

https://github.com/denoland/deno/blob/1464b756a4ef091ef97d1bf7340188ab5fe5492a/cli/tests/unit/fetch_test.ts#L1502-L1508

**Comment by lucacasonato** [comment]
Ah, that makes sense actually:

> A request or response that includes a payload body can include a content-length header field. A request or response is also malformed if the value of a content-length header field does not equal the sum of the DATA frame payload lengths that form the body. A response that is defined to have no payload, as described in [RFC7230], Section 3.3.2, can have a non-zero content-length header field, even though no content is included in DATA frames.

From h2 spec.

**Comment by marcosc90** [comment]
Tests are failing on windows, with the same error as an already ignored test:

https://github.com/denoland/deno/blob/1464b756a4ef091ef97d1bf7340188ab5fe5492a/cli/tests/unit/fetch_test.ts#L1017-L1039

File: ext/fetch/22_body.js [hunk 4]
```diff
@@ -147,14 +151,32 @@
...
           const { value: chunk, done } = await reader.read();
           if (done) break;
-          ArrayPrototypePush(chunks, chunk);
+
+          if (finalBuffer) {
+            // fast path, content-length is present
+            TypedArrayPrototypeSet(finalBuffer, chunk, totalLength);
+          } else {
+            // slow path, content-length is not present
+            // or totalLength > content-length
```
Comment (lucacasonato) [inline ext/fetch/22_body.js:169]:
```suggestion
```

**Review by lucacasonato — APPROVED** [review]
LGTM!
